### PR TITLE
cluster: Scale-out action can be divided into 2

### LINKED ISSUE #1603: TiUP cluster Scale-out action can be divided into 2 steps
## Feature Request

**Is your feature request related to a problem? Please describe:**

Customer installed cluster in version 5.1.1, and they patched ticdc with a hotfix package from a local file.
In this case, when the customer wants to scale out the ticdc , the tiup mirror only has the package of 5.1.1 version. It needs much extra work to package the local patch file and upload to tiup mirror before customer can scale out the ticdc. 

**Describe the feature you'd like:**

We need provide a way to patch the new node during the scale out action.
Current scale out way : 
-- tiup cluster scale-out <cluster-name> scale-out.yaml

Proposed way:
-- 1: Scale-out new nodes with the package in the mirror use a special parameter , then the scale-out action will only deploy the new nodes but not start them.
-- 2:  use patch command to patch the new nodes with local file.
-- 3.   start the new nodes .

**Why the feature is needed:**

Provide a way for user to scale out the ticdc  and patch from local file. 

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by srstack**
### What problem does this PR solve? 
- close #1603 

### What is changed and how it works?

TiUP cluster Scale-out action can be divided into 2

1. tiup cluster scale-out scale-out.yaml --stage1 # deploy
2. tiup cluster scale-out --stage2 # start and reload config
image

When you use stage1 and run successfully, the scale-out file lock will be added, which saves stage1 scale-out topology information.

When the scale-out file lock exists, the following operations cannot be performed:
- tiup cluster start
- tiup cluster restart
- tiup cluster stop
- tiup cluster reload
- tiup cluster scale-out [ --stage1 ]
- tiup cluster scale-in
- tiup cluster upgrade
- tiup cluster patch [--offline=false]

image

If you use --stage1 and --stage2 at the same time, scale-out operations will not be divided into 2
```shell
tiup cluster scale-out --stage1 --stage2 
```
is equivalent to
```shell
tiup cluster scale-out 
```

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation

Release notes:

```release-note
## Added parameters
- stage1
`Don't start the new instance after scale-out, need to manually execute cluster scale-out --stage2`
- stage2
`Start the new instance and init config after scale-out --stage1`
```

### CODE REVIEW

File: components/cluster/command/scale_out.go [hunk 1]
```diff
@@ -30,19 +30,36 @@ func newScaleOutCmd() *cobra.Command {
 		IdentityFile: filepath.Join(utils.UserHome(), ".ssh", "id_rsa"),
 	}
 	cmd := &cobra.Command{
-		Use:          "scale-out <cluster-name> <topology.yaml>",
+		Use:          "scale-out <cluster-name> [topology.yaml]",
```
Comment (nexustar) [inline components/cluster/command/scale_out.go:33]: [] means optional option

File: pkg/cluster/spec/spec_manager.go [hunk 2]
```diff
@@ -202,3 +206,80 @@ func (s *SpecManager) ensureDir(clusterName string) error {
...
+
+// NewScaleOutLock save the meta with specified cluster name.
+func (s *SpecManager) NewScaleOutLock(clusterName string, topo Topology) error {
+	wrapError := func(err error) *errorx.Error {
+		return ErrSaveScaleOutFileFailed.Wrap(err, "Failed to create scale-out file lock")
+	}
+
+	if locked, err := s.IsScaleOutLocked(clusterName); locked {
+		return wrapError(err).
+			WithProperty(tui.SuggestionFromString("The scale out file lock is exist, please run tiup-cluster scale-out --stage2 to continue."))
```
Comment (AstroProfundis) [inline pkg/cluster/spec/spec_manager.go:260]:
```suggestion
			WithProperty(tui.SuggestionFromString("The scale out file lock already exists, please run tiup-cluster scale-out --stage2 to continue."))
```

File: components/cluster/command/scale_out.go [hunk 3]
```diff
@@ -30,19 +30,36 @@ func newScaleOutCmd() *cobra.Command {
...
 		SilenceUsage: true,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if len(args) != 2 {
-				return cmd.Help()
+			var (
+				clusterName string
+				topoFile    string
+			)
+
+			// stage1 and stage2 cannot be true at the same time
```
[the remark this answers is not in the document]
Comment (srstack) [reply components/cluster/command/scale_out.go:42]: If you use --stage1 and --stage2 at the same time, scale-out operations will not be divided into 2
```shell
tiup cluster scale-out --stage1 --stage2 
```
is equivalent to
```shell
tiup cluster scale-out 
```
Or refuse to be true at the same time？

[the remark this answers is not in the document]
Comment (AstroProfundis) [reply components/cluster/command/scale_out.go:42]: I think it's reasonable to allow specifying both args, so maybe just update the comment is fine.

**Review by AstroProfundis — APPROVED** [review]

**Pull request merged**